src/shogun/io/openml/OpenMLRun.cpp

```diff
@@ -0,0 +1,183 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#include <shogun/evaluation/CrossValidationStorage.h>
+#include <shogun/io/openml/OpenMLRun.h>
+#include <shogun/io/openml/ShogunOpenML.h>
+#include <shogun/labels/Labels.h>
+#include <shogun/machine/Machine.h>
+
+using namespace shogun;
+
+std::shared_ptr<OpenMLRun> OpenMLRun::run_model_on_task(
+    std::shared_ptr<CSGObject> model, std::shared_ptr<OpenMLTask> task)
+{
+	SG_SNOTIMPLEMENTED
+	return std::shared_ptr<OpenMLRun>();
+}
+
+std::shared_ptr<OpenMLRun> OpenMLRun::run_flow_on_task(
+    std::shared_ptr<OpenMLFlow> flow, std::shared_ptr<OpenMLTask> task)
+{
+	auto data = task->get_dataset();
+	std::shared_ptr<CFeatures> features = nullptr;
+	std::shared_ptr<CLabels> labels = nullptr;
+
+	auto model = ShogunOpenML::flow_to_model(std::move(flow), true);
+
+	labels = data->get_labels();
+	features = data->get_features(data->get_default_target_attribute());
+
+	auto storage = std::make_shared<CrossValidationStorage>();
+
+	if (task->get_split()->contains_splits())
+	{
+		auto machine = std::dynamic_pointer_cast<CMachine>(model);
+		if (!machine)
+		{
+			SG_SERROR("INTERNAL ERROR: failed to cast model to machine!\n")
+		}
+		auto train_idx = task->get_train_indices();
+		auto test_idx = task->get_test_indices();
+
+		auto xval_storage = std::make_shared<CrossValidationStorage>();
+		xval_storage->set_num_folds(task->get_num_fold());
+		xval_storage->set_num_runs(task->get_num_repeats());
+		machine->set_store_model_features(true);
+
+		// copied/adapted from crossvalidation
+		for (auto repeat_idx : range(task->get_num_repeats()))
+		{
+			for (auto fold_idx : range(task->get_num_fold()))
+			{
+				auto* fold = new CrossValidationFoldStorage();
+				SG_REF(fold)
+
+				auto cloned_machine = (CMachine*)machine->clone();
```

Comment:
this might clone the attached data of the machine. Could we have a method (in develop actually) that clones all model and hyper-parameters of a machine? We somehow need to make a clone without copying the features/labels

Reply:
something like a deep vs shallow clone?

Reply:
kind of, yes. Something temporary is fine (at least better than cloning!)
Eventually, the `view` or `subset` method will return a new instance

I remember @vinx13 was working on this at some point. What is the state of that @vinx13 
Might be something to tackle in Berlin :)

Reply:
We have only features or labels view/subset, but not for machine

Reply:
That’s fine. So view creates a shallow clone and then adds subset? Is that in a branch or in develop? If I’m branch maybe it’s time to merge/finalise it?

Reply:
It creates a shallow clone and adds subset. We already have it in develop branch